### PR TITLE
return more user data if possible for membership tiers

### PR DESCRIPTION
**Pull request by bennycio**
Changes:

- **Changed** update membership tiers background process to store more data for the potential user who owns a membership card such as:

1. Owner Address (will always be available)
2. Username
3. User ID
4. some preview URLs for NFTs they own

### CODE REVIEW

File: server/handler.go [hunk 1]
```diff
@@ -37,7 +37,7 @@ func authHandlersInitToken(parent *gin.RouterGroup, repos *repositories, ethClie
 	usersGroup.POST("/update/addresses/add", middleware.JWTRequired(repos.userRepository, ethClient), addUserAddress(repos.userRepository, repos.nonceRepository, ethClient.EthClient))
 	usersGroup.POST("/update/addresses/remove", middleware.JWTRequired(repos.userRepository, ethClient), removeAddressesToken(repos.userRepository, repos.collectionTokenRepository))
 	usersGroup.GET("/get", middleware.JWTOptional(), getUser(repos.userRepository))
-	usersGroup.GET("/membership", getMembershipTiers(repos.membershipRepository, repos.userRepository, ethClient))
+	usersGroup.GET("/membership", getMembershipTiersToken(repos.membershipRepository, repos.userRepository, repos.tokenRepository, ethClient))
```
Comment (Robinnnnn) [inline server/handler.go:40]: should we rate-limit this endpoint?

File: server/handler.go [hunk 2]
```diff
@@ -60,7 +60,7 @@ func authHandlersInitNFT(parent *gin.RouterGroup, repos *repositories, ethClient
 	usersGroup.POST("/update/addresses/add", middleware.JWTRequired(repos.userRepository, ethClient), addUserAddress(repos.userRepository, repos.nonceRepository, ethClient.EthClient))
 	usersGroup.POST("/update/addresses/remove", middleware.JWTRequired(repos.userRepository, ethClient), removeAddresses(repos.userRepository, repos.collectionRepository))
 	usersGroup.GET("/get", middleware.JWTOptional(), getUser(repos.userRepository))
-	usersGroup.GET("/membership", getMembershipTiers(repos.membershipRepository, repos.userRepository, ethClient))
+	usersGroup.GET("/membership", getMembershipTiers(repos.membershipRepository, repos.userRepository, repos.nftRepository, ethClient))
```
Comment (Robinnnnn) [inline server/handler.go:63]: hmm what's the difference between the group of endpoints in `authHandlersInitNFT` vs `authHandlersInitToken`?

File: server/membership.go [hunk 3]
```diff
@@ -49,101 +46,33 @@ func getMembershipTiers(membershipRepository persist.MembershipRepository, userR
...
-			events, err := openseaFetchMembershipCards(persist.Address(viper.GetString("CONTRACT_ADDRESS")), persist.TokenID(id), 0)
-			if err != nil || len(events) == 0 {
-				tierChan <- tier
-				return
-			}
-			asset := events[0].Asset
-			tier.Name = asset.Name
-			tier.AssetURL = asset.ImageURL
-			tier.Owners = []string{}
+func getMembershipTiersToken(membershipRepository persist.MembershipRepository, userRepository persist.UserRepository, nftRepository persist.TokenRepository, ethClient *eth.Client) gin.HandlerFunc {
```
Comment (Robinnnnn) [inline server/membership.go:49]: same q as before- what's the purpose of this endpoint vs the other?

File: service/membership/membership.go [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+			tier.Owners = make([]persist.MembershipOwner, 0, len(events)*2)
+
+			ownersChan := make(chan []persist.MembershipOwner)
+			for _, e := range events {
+				go func(event opensea.Event) {
+					owners := make([]persist.MembershipOwner, 0, 2)
+					hasNFT, _ := ethClient.HasNFT(pCtx, id, event.FromAccount.Address)
+					if hasNFT {
+						membershipOwner := persist.MembershipOwner{}
+						if glryUser, err := userRepository.GetByAddress(pCtx, event.FromAccount.Address); err == nil && glryUser.UserName != "" {
```
Comment (Robinnnnn) [inline service/membership/membership.go:44]: are we spamming mongo in this for loop? wondering if it's better to use an aggregation query. also curious how long this takes when run against prod DB

File: service/membership/membership.go [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+							}
+						} else {
+							membershipOwner.Address = event.FromAccount.Address
+						}
+						owners = append(owners, membershipOwner)
+					}
+					hasNFT, _ = ethClient.HasNFT(pCtx, id, event.ToAccount.Address)
+					if hasNFT {
+						membershipOwner := persist.MembershipOwner{}
+						if glryUser, err := userRepository.GetByAddress(pCtx, event.ToAccount.Address); err == nil && glryUser.UserName != "" {
```
Comment (Robinnnnn) [inline service/membership/membership.go:75]: same comment

File: service/membership/membership.go [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+		}(v)
+	}
+
+	for i := 0; i < len(middleware.RequiredNFTs); i++ {
+		membershipTiers[i] = <-tierChan
+	}
+	return membershipTiers, nil
+}
+
+// UpdateMembershipTiersToken fetches all membership cards for a token ID
```
Comment (Robinnnnn) [inline service/membership/membership.go:120]: this has the same description as the previous func 🤔

File: service/membership/membership.go [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+	}
+
+	response := &opensea.Events{}
+	err = util.UnmarshallBody(response, resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	result = append(result, response.Events...)
+	if len(response.Events) == 100 {
+		next, err := openseaFetchMembershipCards(contractAddress, tokenID, pOffset+100)
```
Comment (Robinnnnn) [inline service/membership/membership.go:220]: what are the odds of getting throttled by opensea? should we add a small timeout?

**Review by kaitoo1 — APPROVED** [review]
LGTM, besides what Robin mentioned about getting throttled by OS

**Pull request merged**